src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,544 @@
+            Entry[] newEntries = new Entry[newSize];
+
+            int count = Count;
+            Array.Copy(_entries, newEntries, count);
+            for (int i = 0; i < count; ++i)
+            {
+                ref Entry entry = ref newEntries[i];
+                ref int next = ref newBuckets[entry.HashCode % newSize];
+                entry.Next = next - 1;
+                next = i + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = newEntries;
+        }
+
+        private int IndexOf(TKey key, out int hashCode)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            IEqualityComparer<TKey> comparer = _comparer;
+            hashCode = (comparer?.GetHashCode(key) ?? key.GetHashCode()) & 0x7FFFFFFF;
+            int[] buckets = _buckets;
+            int index = -1;
+            if (buckets.Length > 0)
+            {
+                int bucket = hashCode % buckets.Length;
+                index = buckets[bucket] - 1;
+                comparer = comparer ?? EqualityComparer<TKey>.Default;
+                Entry[] entries = _entries;
+                while (index >= 0)
+                {
+                    Entry entry = entries[index];
+                    if (entry.HashCode == hashCode && comparer.Equals(entry.Key, key))
+                    {
+                        break;
+                    }
+                    index = entry.Next;
+                }
+            }
+            return index;
+        }
+
+        private bool TryInsert(int? index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            int i = IndexOf(key, out int hashCode);
+            Entry[] entries = _entries;
+            if (i >= 0)
+            {
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        entries[i].Value = value;
+                        return true;
+                    case InsertionBehavior.ThrowOnExisting:
+                        throw new ArgumentException(string.Format(Strings.Argument_AddingDuplicateWithKey, key.ToString()));
+                    default:
+                        return false;
+                }
+            }
+
+            ++_version;
+            // Check if resize is needed
+            int count = Count;
+            if (entries.Length == count)
+            {
+                Resize(HashHelpers.ExpandPrime(entries.Length));
+                entries = _entries;
+            }
+
+            // Increment indices >= index;
+            int actualIndex = index ?? count;
+            for (i = actualIndex; i < count; ++i)
+            {
+                UpdateBucketIndices(i, 1);
+            }
+            Array.Copy(entries, actualIndex, entries, actualIndex + 1, count - actualIndex);
+
+            ref int bucket = ref _buckets[hashCode % _buckets.Length];
```

Comment:
Per the perf measurements in DictionarySlim, it might be faster to just store `hashCode % _buckets.Length` and index into `_buckets` again when you want to set it. Perhaps you already tried though.

Reply:
Same comment other places. Only measurements can say.

Reply:
I'll leave it as is until performance tests are added.